// @pulumi/cloud-aws/shared.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
const aws = require("@pulumi/aws");
const pulumi = require("@pulumi/pulumi");
const errors_1 = require("@pulumi/pulumi/errors");
const config = require("./config");
const clusterMod = require("./cluster");
const networkMod = require("./network");
// nameWithStackInfo is the resource prefix we'll use for all resources we auto-provision.  In general,
// it's safe to use these for top-level components like Network and Cluster, because they suffix all
// internal resources they provision.
const nameWithStackInfo = `pulumi-${pulumi.getStack()}`;
function createNameWithStackInfo(requiredInfo) {
    const maxLength = 24;
    if (requiredInfo.length > maxLength) {
        throw new errors_1.RunError(`'${requiredInfo}' cannot be longer then ${maxLength} characters.`);
    }
    // No required portion.  Just return the stack name.
    if (requiredInfo.length === 0) {
        return nameWithStackInfo.substr(0, maxLength);
    }
    // Only enough room for required portion, don't add the stack.
    // Also don't add the stack if there wouldn't be room to add it and a dash.
    if (requiredInfo.length >= maxLength - "-".length) {
        return requiredInfo;
    }
    // Attempt to keep some portion of the stack, then - then the required part.
    const suffix = "-" + requiredInfo;
    const result = nameWithStackInfo.substr(0, maxLength - suffix.length) + suffix;
    return result;
}
exports.createNameWithStackInfo = createNameWithStackInfo;
// Expose a common infrastructure resource that all our global resources can consider themselves to
// be parented by.  This helps ensure unique URN naming for these guys as tey cannot conflict with
// any other user resource.
class InfrastructureResource extends pulumi.ComponentResource {
    constructor() {
        super("cloud:global:infrastructure", "global-infrastructure");
        this.registerOutputs();
    }
}
let globalInfrastructureResource;
function getGlobalInfrastructureResource() {
    if (!globalInfrastructureResource) {
        globalInfrastructureResource = new InfrastructureResource();
    }
    return globalInfrastructureResource;
}
exports.getGlobalInfrastructureResource = getGlobalInfrastructureResource;
// Whether or not we should run lamabda-based compute in the private network
exports.runLambdaInVPC = config.usePrivateNetwork;
// The IAM Role Policies to apply to compute for both Lambda and ECS
const defaultComputePolicies = [
    aws.iam.AWSLambdaFullAccess,
    aws.iam.AmazonEC2ContainerServiceFullAccess,
];
let computePolicies = config.computeIAMRolePolicyARNs
    ? config.computeIAMRolePolicyARNs.split(",")
    : defaultComputePolicies;
let computePoliciesAccessed = false;
// Set the IAM policies to use for compute.
function setComputeIAMRolePolicies(policyARNs) {
    if (computePoliciesAccessed) {
        throw new errors_1.RunError("The compute policies have already been used, make sure you are setting IAM policies early enough.");
    }
    computePolicies = policyARNs;
}
exports.setComputeIAMRolePolicies = setComputeIAMRolePolicies;
// Get the IAM policies to use for compute.
function getComputeIAMRolePolicies() {
    computePoliciesAccessed = true;
    return computePolicies;
}
exports.getComputeIAMRolePolicies = getComputeIAMRolePolicies;
let network;
/**
 * Get or create the network to use for container and lambda compute.
 */
function getOrCreateNetwork() {
    if (!network) {
        if (!config.externalVpcId) {
            if (config.usePrivateNetwork) {
                // Create a new VPC for this private network.
                network = new networkMod.Network(createNameWithStackInfo("global"), {
                    usePrivateSubnets: config.usePrivateNetwork,
                    numberOfAvailabilityZones: config.ecsAutoCluster ? config.ecsAutoClusterNumberOfAZs : undefined,
                });
            }
            else {
                // Use the default VPC.
                network = networkMod.Network.getDefault();
            }
        }
        else /* config.externalVpcId */ {
            if (!config.externalSubnets || !config.externalSecurityGroups || !config.externalPublicSubnets) {
                throw new errors_1.RunError("If providing 'externalVpcId', must provide 'externalSubnets', " +
                    "'externalPublicSubnets' and 'externalSecurityGroups'");
            }
            // Use an exsting VPC for this private network
            network = networkMod.Network.fromVpc("external-vpc", {
                vpcId: config.externalVpcId,
                usePrivateSubnets: config.usePrivateNetwork,
                subnetIds: config.externalSubnets,
                publicSubnetIds: config.externalPublicSubnets,
                securityGroupIds: config.externalSecurityGroups,
            });
        }
    }
    return network;
}
exports.getOrCreateNetwork = getOrCreateNetwork;
/**
 * @deprecated
 */
function getNetwork() {
    return getOrCreateNetwork();
}
exports.getNetwork = getNetwork;
// The cluster to use for container compute or undefined if containers are unsupported.
let cluster;
function getCluster() {
    // If no ECS cluster has been initialized, see if we must lazily allocate one.
    if (!cluster) {
        if (config.ecsAutoCluster) {
            // Translate the comma-seperated list into an array or undefined.
            let instanceRolePolicyARNs;
            if (config.ecsAutoClusterInstanceRolePolicyARNs) {
                instanceRolePolicyARNs = (config.ecsAutoClusterInstanceRolePolicyARNs || "").split(",");
            }
            // If we are asked to provision a cluster, then we will have created a network
            // above - create a cluster in that network.
            cluster = new clusterMod.Cluster(createNameWithStackInfo("global"), {
                network: getOrCreateNetwork(),
                addEFS: config.ecsAutoClusterUseEFS,
                instanceType: config.ecsAutoClusterInstanceType,
                instanceRolePolicyARNs: instanceRolePolicyARNs,
                instanceRootVolumeSize: config.ecsAutoClusterInstanceRootVolumeSize,
                instanceDockerImageVolumeSize: config.ecsAutoClusterInstanceDockerImageVolumeSize,
                instanceSwapVolumeSize: config.ecsAutoClusterInstanceSwapVolumeSize,
                minSize: config.ecsAutoClusterMinSize,
                maxSize: config.ecsAutoClusterMaxSize,
                publicKey: config.ecsAutoClusterPublicKey,
            });
        }
        else if (config.ecsClusterARN) {
            // Else if we have an externally provided cluster and can use that.
            cluster = {
                ecsClusterARN: pulumi.output(config.ecsClusterARN),
                securityGroupId: config.ecsClusterSecurityGroup
                    ? pulumi.output(config.ecsClusterSecurityGroup) : undefined,
                efsMountPath: config.ecsClusterEfsMountPath,
            };
        }
        else if (config.useFargate) {
            // Else, allocate a Fargate-only cluster.
            cluster = new clusterMod.Cluster(createNameWithStackInfo("global"), {
                network: getOrCreateNetwork(),
                maxSize: 0,
                addEFS: false,
            });
        }
    }
    return cluster;
}
exports.getCluster = getCluster;
